rust/zephyr/src/gpio.rs

```diff
@@ -0,0 +1,950 @@
+                    C::gpio_pin_interrupt_configure(
+                        self.0 as *const _ as *mut _,
+                        p,
+                        interrupt_config,
+                    )?;
+                }
+            }
+            self.1 &= !mask;
+            Ok(InterruptPinGroup(InputPinGroup(self.0, mask)))
+        } else {
+            Err(io::Error::from_raw_os_error(ERR))
+        }
+    }
+
+    pub fn split_into_output<C: GpioSyscalls>(
+        &mut self,
+        mask: Pins,
+        config: OutputConfig,
+        flags: Option<ConfigFlags>,
+    ) -> io::Result<OutputPinGroup> {
+        const ERR: i32 = -(raw::EINVAL as i32);
+
+        if self.1 & mask == mask {
+            for p in mask.iter() {
+                unsafe {
+                    C::gpio_config(self.0 as *const _ as *mut _, p, config.into(), flags)?;
+                }
+            }
+            self.1 &= !mask;
+            Ok(OutputPinGroup(self.0, mask))
+        } else {
+            Err(io::Error::from_raw_os_error(ERR))
+        }
+    }
+}
+
+pub struct InputPinGroup(&'static Device, Pins);
+impl PinGroup for InputPinGroup {
+    fn get_mask(&self) -> Pins {
+        self.1
+    }
+}
+impl InputPinGroup {
+    /// # Safety
+    ///
+    /// The caller must ensure both arguments are using the same port.
+    #[inline(always)]
+    pub unsafe fn merge<T: PinGroup>(&mut self, pins: InputPinGroup) {
+        self.1 |= pins.get_mask();
+    }
+
+    #[inline(always)]
+    pub fn split(mut self, pins: Pins) -> Option<(InputPinGroup, InputPinGroup)> {
+        if self.1 & pins == pins {
+            self.1 &= !pins;
+            Some((InputPinGroup(self.0, pins), self))
+        } else {
+            None
+        }
+    }
+
+    pub fn into_dynamic<C: GpioSyscalls>(self) -> io::Result<DynamicPinGroup> {
+        let mut d = DynamicPinGroup(self.0, Pins::NONE);
+        d.merge::<C, _>(self)?;
+        Ok(d)
+    }
+
+    pub fn into_interrupt<C: GpioSyscalls>(
+        self,
+        config: InterruptConfig,
+    ) -> io::Result<InterruptPinGroup> {
+        for p in self.1.iter() {
+            unsafe {
+                C::gpio_pin_interrupt_configure(self.0 as *const _ as *mut _, p, config)?;
+            }
+        }
+        Ok(InterruptPinGroup(self))
+    }
+
+    #[inline(always)]
+    fn port_get_raw<C: GpioSyscalls>(&self) -> io::Result<Pins> {
+        unsafe { C::gpio_port_get_raw(self.0 as *const _ as *mut _) }
+    }
+
+    #[inline(always)]
+    fn port_get<C: GpioSyscalls>(&self) -> io::Result<Pins> {
+        const ERR: i32 = -(raw::ENOTSUP as i32);
+
+        let value = self.port_get_raw::<C>()?;
+
+        let invert = if self.0.data.is_null() {
+            Err(io::Error::from_raw_os_error(ERR))
+        } else {
+            let driver_data: *const raw::gpio_driver_data =
+                self.0.data as *const raw::gpio_driver_data;
+            Ok(unsafe { (*driver_data).invert })
```

Comment:
I get a compile error on Zephyr 2.3 because the device private data field is named differently.

```
763 |                 self.0.data as *const raw::gpio_driver_data;                                                                                                                                                                                                                        
    |                        ^^^^ unknown field                                                                                                                                                                                                                                           
    |                                                                                                                                                                                                                                                                                     
    = note: available fields are: `name`, `config_info`, `driver_api`, `driver_data`
```
I assume the inversion is to account for active low pins. We should use `gpio_port_get` instead of the `_raw` versions so we're not dependent on internal data structures. I see that probably wasn't possible because those functions are inline and not syscalls, so they don't get bindings.

We can create our own thunks for functions that need to be real symbols to supplement the ones automatically generate for syscalls. If you look at the generated file here, you can see how those are defined to emit a real function for each context.

`./build/zephyr/include/generated/syscall_thunks.c`

```
#if defined(__ZEPHYR_SUPERVISOR__)                                                                                                                                                                                                                                                        
int z_kernelctx_gpio_port_get_raw(struct device * port, gpio_port_value_t * value)                                                                                                                                                                                                        
#elif defined(__ZEPHYR_USER__)                                                                                                                                                                                                                                                            
int z_userctx_gpio_port_get_raw(struct device * port, gpio_port_value_t * value)                                                                                                                                                                                                          
#else                                                                                                                                                                                                                                                                                     
int z_anyctx_gpio_port_get_raw(struct device * port, gpio_port_value_t * value)                                                                                                                                                                                                           
#endif                                                                                                                                                                                                                                                                                    
{                                                                                                                                                                                                                                                                                         
        return gpio_port_get_raw(port, value);                                                                                                                                                                                                                                            
}
```

We could manually create something like the above for the missing gpio functions, maybe call it `gpio_thunks.c` and include it from `syscall-thunk-{any,user,kernel}.c`